File: hackerrank/weekofcode28/ValueOfFriendship.java
package hackerrank.weekofcode28;
import java.io.*;
import java.util.*;

public class ValueOfFriendship {
	void run() {
		int n = in.nextInt();
		int m = in.nextInt();
		DisjointSetUnion dsu = new DisjointSetUnion(n);
		int[] a = new int[m];
		int[] b = new int[m];
		for (int i = 0; i < m; i++) {
			a[i] = in.nextInt() - 1; 
			b[i] = in.nextInt() - 1;
			dsu.unite(a[i], b[i]);
		}
		long[] size = new long[n];
		for (int i = 0; i < n; i++) {
			size[dsu.get(i)] += m + 1;
		}
		for (int i = 0; i < m; i++) {
			size[dsu.get(a[i])]++;
		}
		Arrays.sort(size);
		long ans = 0;
		long sum = 0;
		for (int j = n - 1; j >= 0; j--) {
			long s = size[j];
			int v = (int) (s / (m + 1));
			if (v == 0) {
				continue;
			}
			for (int i = 2; i <= v; i++) {
				long cur = i * (i - 1L);
				ans += sum + cur;
			}
			sum += v * (v - 1L);
		}
		for (int j = n - 1; j >= 0; j--) {
			long s = size[j];
			int v = (int) (s / (m + 1));
			if (v == 0) {
				continue;
			}
			int e = (int) (s % (m + 1));
			ans += sum * (e - v + 1);
		}
		out.println(ans);
	}

	static class DisjointSetUnion {
		int[] p;
		Random r = new Random(1);
		
		public DisjointSetUnion(int n) {
			p = new int[n];
			clear();
		}
		
		void clear() {
			for (int i = 0; i < p.length; i++) {
				p[i] = i;
			}
		}

		int get(int v) {
			if (p[v] == v) {
				return v;
			}
			p[v] = get(p[v]);
			return p[v];
		}
		
		void unite(int v, int u) {
			v = get(v);
			u = get(u);
			if (r.nextBoolean()) {
				p[v] = u;
			} else {
				p[u] = v;
			}
		}
	}
	
	static MyScanner in;
	static PrintWriter out;

	public static void main(String[] args) throws IOException {
		boolean stdStreams = true;
		String fileName = ValueOfFriendship.class.getSimpleName().replaceFirst("_.*", "").toLowerCase();
		String inputFileName = fileName + ".in";
		String outputFileName = fileName + ".out";
		
		Locale.setDefault(Locale.US);
		BufferedReader br;
		if (stdStreams) {
			br = new BufferedReader(new InputStreamReader(System.in));
			out = new PrintWriter(System.out);
		} else {
			br = new BufferedReader(new FileReader(inputFileName));
			out = new PrintWriter(outputFileName);
		}
		in = new MyScanner(br);
		int tests = in.nextInt();
		for (int test = 0; test < tests; test++) {
			new ValueOfFriendship().run();
		}
		br.close();
		out.close();
	}
	
	static class MyScanner {
		BufferedReader br;
		StringTokenizer st;

		MyScanner(BufferedReader br) {
			this.br = br;
		}
		
		void findToken() {
			while (st == null || !st.hasMoreTokens()) {
				try {
					st = new StringTokenizer(br.readLine());
				} catch (IOException e) {
					throw new RuntimeException(e);
				}
			}
		}
		
		String next() {
			findToken();
			return st.nextToken();
		}
		
		int nextInt() {
			return Integer.parseInt(next());
		}
		
		long nextLong() {
			return Long.parseLong(next());
		}
		
		double nextDouble() {
			return Double.parseDouble(next());
		}
	}
}

File: hackerrank/weekofcode28/value-of-friendship.py
#!/usr/bin/env python3
import random

def solve(n, edges):
	dsu = Dsu(n)
	for pair in edges:
		dsu.unite(*pair)
	size = [[0, 0] for _ in range(n)]
	for i in range(n):
		size[dsu.get(i)][0] += 1
	for pair in edges:
		size[dsu.get(pair[0])][1] += 1
	size = [(v, e) for (v, e) in size if v > 0]
	size.sort(reverse=True)
	ans = 0
	sum = 0
	for v, e in size:
		cur = 0
		for i in range(2, v + 1):
			cur = i * (i - 1)
			ans += sum + cur
		sum += cur
	for v, e in size:
		ans += (e - v + 1) * sum
	return ans

class Dsu:
	def __init__(self, n):
		self.r = random.Random(1)
		self.p = [i for i in range(n)]

	def get(self, v):
		if self.p[v] == v:
			return v
		self.p[v] = self.get(self.p[v])
		return self.p[v]

	def unite(self, v, u):
		v, u = self.get(v), self.get(u)
		if self.r.getrandbits(1):
			v, u = u, v
		self.p[v] = u

def test():
	assert solve(5, [(0, 1), (2, 1), (3, 1), (3, 2)]) == 32

def read_test():
	n, m = map(int, input().split())
	edges = [tuple([int(s) - 1 for s in input().split()]) for _ in range(m)]
	return (n, edges)


if __name__ == '__main__':
	#test()
	for _ in range(int(input())):
		print(solve(*read_test()))
